polar-core/src/partial/constraints.rs

```diff
@@ -0,0 +1,501 @@
+use serde::{Deserialize, Serialize};
+
+use crate::counter::Counter;
+use crate::error::{OperationalError, PolarResult};
+use crate::events::QueryEvent;
+use crate::runnable::Runnable;
+use crate::terms::{Operation, Operator, Pattern, Symbol, Term, Value};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct Constraints {
+    operations: Vec<Operation>,
+    variable: Symbol,
+}
+
+impl Constraints {
+    pub fn new(variable: Symbol) -> Self {
+        Constraints {
+            operations: vec![],
+            variable,
+        }
+    }
+
+    pub fn operations(&self) -> &Vec<Operation> {
+        &self.operations
+    }
+
+    pub fn operations_mut(&mut self) -> &mut Vec<Operation> {
+        &mut self.operations
+    }
+
+    pub fn unify(&mut self, other: Term) {
+        let op = op!(Unify, self.variable_term(), other);
+        self.operations.push(op);
+    }
+
+    pub fn isa(&mut self, other: Term) -> Box<dyn Runnable> {
+        let isa_op = op!(Isa, self.variable_term(), other);
+
+        let constraint_check = Box::new(IsaConstraintCheck::new(
+            self.operations.clone(),
+            isa_op.clone(),
+        ));
+
+        self.operations.push(isa_op);
+        constraint_check
+    }
+
+    pub fn compare(&mut self, operator: Operator, other: Term) {
+        assert!(matches!(
+            operator,
+            Operator::Lt
+                | Operator::Gt
+                | Operator::Leq
+                | Operator::Geq
+                | Operator::Eq
+                | Operator::Neq
+        ));
+
+        let op = Operation {
+            operator,
+            args: vec![self.variable_term(), other],
+        };
+
+        self.operations.push(op);
+    }
+
+    /// Add lookup of `field` assigned to `value` on `self.
+    ///
+    /// Returns: A partial expression for `value`.
+    pub fn lookup(&mut self, field: Term, value: Term) -> Term {
+        // Note this is a 2-arg lookup (Dot) not 3-arg. (Pre rewrite).
+        assert!(matches!(field.value(), Value::String(_)));
+
+        self.operations.push(op!(
+            Unify,
+            value.clone(),
+            term!(op!(Dot, self.variable_term(), field))
+        ));
+
+        let name = value.value().as_symbol().unwrap();
+        Term::new_temporary(Value::Partial(Constraints::new(name.clone())))
```

Comment:
This assumes that `value` is a variable, which I believe is guaranteed due to how the 2 => 3 arg rewrite works.

In the example you brought up, the unify would be applied to the temporary variable.  Recall that `partial1.field = value1` gets rewritten into `partial1.field = _temp_1 and _temp_1 = value`.

The returned partial is bound to `_temp_1` in the VM, so that when `_temp_1 = value` is evaluated, the constraint is added there.

Does this make sense to you?

Reply:
Oh yeah, I forgot about rewrites.